src/VisualStudio/Core/Def/Implementation/DesignerAttribute/DesignerAttributeIncrementalAnalyzer.cs

```diff
@@ -134,28 +131,41 @@ private async Task<DesignerAttributeResult> ScanDesignerAttributesOnRemoteHostIf
             return await service.ScanDesignerAttributesAsync(document, cancellationToken).ConfigureAwait(false);
         }
 
-        private async Task<bool> IsCpsProjectAsync(Project project, CancellationToken cancellationToken)
+        private async Task<IProjectItemDesignerTypeUpdateService> GetUpdateServiceIfCpsProjectAsync(Project project, CancellationToken cancellationToken)
         {
             if (_cpsProjects.TryGetValue(project.Id, out var value))
             {
                 return value;
             }
 
-            // CPS projects do not support designer attributes.  So we just skip these projects entirely.
             var vsWorkspace = project.Solution.Workspace as VisualStudioWorkspaceImpl;
 
             await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(alwaysYield: true, cancellationToken);
             cancellationToken.ThrowIfCancellationRequested();
 
-            var cps = vsWorkspace?.IsCPSProject(project) == true;
+            if (!vsWorkspace.IsCPSProject(project))
+            {
+                _cpsProjects.TryAdd(project.Id, null);
+                return null;
+            }
 
-            // The remainder of this method does not need to execute on the UI thread, but it's pointless to force a
-            // context switch if the caller of this method is the UI thread.
+            var vsProject = (IVsProject)vsWorkspace.GetHierarchy(project.Id);
+            if (ErrorHandler.Failed(vsProject.GetItemContext((uint)VSConstants.VSITEMID.Root, out var projectServiceProvider)))
+            {
+                _cpsProjects.TryAdd(project.Id, null);
+                return null;
+            }
 
-            _cpsProjects.TryAdd(project.Id, cps);
+            var serviceProvider = new Shell.ServiceProvider(projectServiceProvider);
+            var updateService = serviceProvider.GetService(typeof(IProjectItemDesignerTypeUpdateService)) as IProjectItemDesignerTypeUpdateService;
```

Comment:
We will make sure it is free threaded to get this service inside CPS project.  (We have a pending insertion to the 16.1 update preview 2, and we should try to remove 'SwitchToUIThread' after that point.

Reply:
okay.